packages/webdriverio/src/utils.js

```diff
@@ -9,135 +9,182 @@ import isObject from 'lodash.isobject'
 import { ELEMENT_KEY, W3C_SELECTOR_STRATEGIES, UNICODE_CHARACTERS } from './constants'
 
 const log = logger('webdriverio')
-const DEFAULT_SELECTOR = 'css selector'
+const DEFAULT_STRATEGY = 'css selector'
 const DIRECT_SELECTOR_REGEXP = /^(id|css selector|xpath|link text|partial link text|name|tag name|class name|-android uiautomator|-ios uiautomation|-ios predicate string|-ios class chain|accessibility id):(.+)/
 const INVALID_SELECTOR_ERROR = new Error('selector needs to be typeof `string` or `function`')
 
-export const findStrategy = function (value, isW3C, isMobile) {
-    const isNameAttribute = value.search(/^\[name=("|')([a-zA-z0-9\-_. ]+)("|')]$/) >= 0
-
+const defineStrategy = function (selector) {
     /**
-     * set default selector
+     * check if user has specified locator strategy directly
      */
-    let using = DEFAULT_SELECTOR
+    if (selector.match(DIRECT_SELECTOR_REGEXP)) {
+        return 'directly'
+    }
+    // use xPath strategy if selector starts with //
+    if (selector.indexOf('/') === 0 || selector.indexOf('(') === 0 ||
+        selector.indexOf('../') === 0 || selector.indexOf('./') === 0 ||
+        selector.indexOf('*/') === 0) {
+        return 'xpath'
+    }
+    // use link text strategy if selector starts with =
+    if (selector.indexOf('=') === 0) {
+        return 'link text'
+    }
+    // use partial link text strategy if selector starts with *=
+    if (selector.indexOf('*=') === 0) {
+        return 'partial link text'
+    }
+    // recursive element search using the UiAutomator library (Android only)
+    if (selector.indexOf('android=') === 0) {
+        return '-android uiautomator'
+    }
+    // recursive element search using the UIAutomation library (iOS-only)
+    if (selector.indexOf('ios=') === 0) {
+        return '-ios uiautomation'
+    }
+    // recursive element search using accessibility id
+    if (selector.indexOf('~') === 0) {
+        return 'accessibility id'
+    }
+    // class name mobile selector
+    // for iOS = UIA...
+    // for Android = android.widget
+    if (selector.slice(0, 3) === 'UIA' || selector.slice(0, 15) === 'XCUIElementType' || selector.slice(0, 14).toLowerCase() === 'android.widget') {
+        return 'class name'
+    }
+    // use tag name strategy if selector contains a tag
+    // e.g. "<div>" or "<div />"
+    if (selector.search(/<[a-zA-Z-]+( \/)*>/g) >= 0) {
+        return 'tag name'
+    }
+    // use name strategy if selector queries elements with name attributes for JSONWP
+    // or if isMobile is used even when w3c is used
+    // e.g. "[name='myName']" or '[name="myName"]'
+    if (selector.search(/^\[name=("|')([a-zA-z0-9\-_. ]+)("|')]$/) >= 0) {
+        return 'name'
+    }
+    // allow to move up to the parent or select current element
+    if (selector === '..' || selector === '.') {
```

Comment:
can it be combined with xpath?

Reply:
yes it can be, but I don't want to. Because this two conditions have different meaning and I agree with author to split it into two different conditions.
So we have:
1. `use xPath strategy if selector starts with //`
```js
if (selector.indexOf('/') === 0 || selector.indexOf('(') === 0 ||
        selector.indexOf('../') === 0 || selector.indexOf('./') === 0 ||
        selector.indexOf('*/') === 0)
```

2. `allow to move up to the parent or select current element`
```js
if (selector === '..' || selector === '.')
```

The nice part that we need one place for that in `findStrategy` method:
```js
    case 'xpath': {
        using = 'xpath'
        break
    }
```